### PR TITLE
OP-377 Implements JWT authentication

### PR DESCRIPTION
**Pull request by npasquetto**
The purpose of this PR is to replace the *Basic* Authentication mechanism with the *JWT* Authentication mechanism as a stateless, more secure, cookie independant solution.

**Spring Security**

Fully integrated in Spring Security, main components are:
* a **token provider** that generate the token
* a **login success handler** thet return the generated token on succesfully login
* a **security filter** that check and validate the token

**SpringFox**

* updated spring fox version (that need an update on spring boot and the addition of gson dependency)
* updated *SpringFoxConfig* class to support the JWT security schema and context

### CODE REVIEW

File: src/main/java/org/isf/security/LoginResponse.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.isf.security;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/LoginResponse.java:1]: Missing license header.

File: src/main/java/org/isf/security/OHSimpleUrlAuthenticationSuccessHandler.java [hunk 2]
```diff
@@ -27,6 +27,8 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/OHSimpleUrlAuthenticationSuccessHandler.java:30]: per project import ordering standards this import should appear after the `org` ones

File: src/main/java/org/isf/security/jwt/JWTConfigurer.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package org.isf.security.jwt;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/jwt/JWTConfigurer.java:1]: Missing license header

File: src/main/java/org/isf/security/jwt/JWTFilter.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package org.isf.security.jwt;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/jwt/JWTFilter.java:1]: missing license header

File: src/main/java/org/isf/security/jwt/JWTFilter.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.GenericFilterBean;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/jwt/JWTFilter.java:13]: imports in wrong order; see in this repo `.ide-settings\eclipse\OpenHosptital,importorder`.   The order is:

1. static
2. java
3. javax
4. org
5. com
6. everything else

File: src/main/java/org/isf/security/jwt/TokenProvider.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
+package org.isf.security.jwt;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/jwt/TokenProvider.java:1]: Same two comments as above.

**Comment by npasquetto** [comment]
Done some fixes about license and import order.

**Comment by mwithi** [comment]
Very nice! Nevertheless, while I'm able to get and use the generated token from Postman with same CURL command, I'm not able to get it working in Swagger: login/ returns response code 0, and although I set it as authorization key, every api return same response. 

Same for you?

File: pom.xml [hunk 7]
```diff
@@ -103,7 +103,12 @@
         <dependency>
             <groupId>io.springfox</groupId>
             <artifactId>springfox-swagger2</artifactId>
-            <version>2.6.1</version>
+            <version>2.9.2</version>
+        </dependency>
+        <dependency>
```
Comment (dbmalkovsky) [inline pom.xml:108]: I don't see were this dependency is imported; is it really used?

Comment (mwithi) [reply pom.xml:108]: @npasquetto do we?

Comment (dbmalkovsky) [reply pom.xml:108]: Just to be clear I was asking about the google.guava dependency at lines 109-110

Comment (mwithi) [reply pom.xml:108]: good point, I understood springfox-swagger2 indeed 😆

**Comment by npasquetto** [comment]
> Very nice! Nevertheless, while I'm able to get and use the generated token with Postman using same CURL command, I'm not able to get it work in Swagger: `login/` returns `response code: 0`, and although I set it as authorization key, every `api/` using the token in the CURL command, return same response.
> 
> Same for you?

Yes, it was due to an error in `api.host` default value in SpringFox configuration, now it should to be ok.
Thanks @mwithi!

File: pom.xml (hunk 7 above)
Comment (npasquetto) [reply pom.xml:108]: No: I've introduced google.guava dependency as a workaround for a bug in a previous version of springfox2 (https://github.com/springfox/springfox/issues/2232) that I've tested before the MR submission, but in the springfox2 version used in the MR, the bug it's been fixed, so the workaround (and the dependency) it's no longer needed.

**Comment by mwithi** [comment]
> Yes, it was due to an error in `api.host` default value in SpringFox configuration, now it should to be ok.
> Thanks @mwithi!

Thanks a lot @npasquetto !!
Now the login is working fine in Swagger, but the token seems not been recognized using Swagger (with Postman the same token do work). 

Maybe I'm doing wrong?

image

image

**Comment by npasquetto** [comment]
> > Yes, it was due to an error in `api.host` default value in SpringFox configuration, now it should to be ok.
> > Thanks @mwithi!
> 
> Thanks a lot @npasquetto !!
> Now the login is working fine in Swagger, but the token seems not been recognized using Swagger (with Postman the same token do work). 
> 
> Maybe I'm doing wrong?
> 
> image
> 
> image
> 
> 
> 
> 

In this version of swagger-ui you need to prepend the token with the 

'Bearer ' 

keyword (space included); this is needed by authorization filters to distinguish between credentials type - basic, bearer, others. 

Check out how Postman set the `Authorization` header on outgoing request.

BTW, I'm going to submit a PR to update the swagger-ui version, as in newer versions this behavior is managed.

**Comment by mwithi** [comment]
> keyword (space included); this is needed by authorization filters to distinguish between credentials type - basic, bearer, others.

I was just wondering how it could guess it! Thanks, it works perfectly now!

> BTW, I'm going to submit a PR to update the swagger-ui version, as in newer versions this behavior is managed.

If I'm not wrong we were waiting for bumping this version (there's a dependabot PR about) because some legacy reasons due to Spring 1.x

**Pull request merged**